### PR TITLE
[Impeller] Implement drawPoints in Impeller

### PR DESCRIPTION
**Pull request by flar**
It may seem like this is a broken implementation calling drawLine on individual paths, but that is actually the semantics of the call. We could do these in parallel for performance, but they must behave as if they are individual calls to drawLine - complete with Polygon mode not connecting with joins and all of the individual lines/points/polygon segments over-drawing at their overlap points. The following test case helps show the behavior and shows the same behavior with and without Impeller.

<details>
<summary>DrawPoints interactive app</summary>

```
import 'dart:typed_data';
import 'dart:math' as math;
import 'dart:ui' as ui;

import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class MobilePoint {
  static math.Random generator = math.Random();

  MobilePoint.random(this.size) {
    x = generator.nextDouble() * size.width;
    y = generator.nextDouble() * size.height;
    dx = (generator.nextDouble() - 0.5) * size.width / 100;
    dy = (generator.nextDouble() - 0.5) * size.height / 100;
  }

  void bump() {
    x += dx;
    if (x > size.width) {
      x = size.width * 2 - x;
      dx = -dx;
    } else if (x < 0) {
      x = -x;
      dx = -dx;
    }
    y += dy;
    if (y > size.height) {
      y = size.height * 2 - y;
      dy = -dy;
    } else if (y < 0) {
      y = -y;
      dy = -dy;
    }
  }

  Size size;
  double x = 0, y = 0;
  double dx = 0, dy = 0;
}

class _MyHomePageState extends State<MyHomePage> with SingleTickerProviderStateMixin {
  static Size pointFieldSize = Size(200, 200);
  List<MobilePoint> points = List<MobilePoint>.generate(100, (index) => MobilePoint.random(pointFieldSize));
  late AnimationController controller;

  @override
  void initState() {
    super.initState();
    controller = AnimationController(vsync: this);
    controller.repeat(period: Duration(seconds: 10));
  }

  @override
  void dispose() {
    controller.stop();
    controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: AnimatedBuilder(
          animation: controller,
          builder: (context, child) {
            final Float32List pointList = Float32List(points.length * 2);
            for (int i = 0; i < points.length; i++) {
              points[i].bump();
              pointList[i * 2] = points[i].x;
              pointList[i * 2 + 1] = points[i].y;
            }
            return Column(
              mainAxisAlignment: MainAxisAlignment.spaceEvenly,
              crossAxisAlignment: CrossAxisAlignment.center,
              children: <Widget>[
                Row(
                  mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                  children: <Widget>[
                    CustomPaint(
                      size: pointFieldSize,
                      painter: PointPainter(pointList, ui.PointMode.points),
                    ),
                    CustomPaint(
                      size: pointFieldSize,
                      painter: PointPainter(pointList, ui.PointMode.lines),
                    ),
                  ],
                ),
                Row(
                  mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                  children: <Widget>[
                    CustomPaint(
                      size: pointFieldSize,
                      painter: PointPainter(pointList, ui.PointMode.polygon),
                    ),
                    CustomPaint(
                      size: pointFieldSize,
                      painter: PathPainter(points),
                    ),
                  ],
                ),
              ],
            );
          },
        ),
      ),
    );
  }
}

class PointPainter extends CustomPainter {
  PointPainter(this.list, this.mode);

  final ui.PointMode mode;
  final Float32List list;

  @override
  void paint(Canvas canvas, Size size) {
    Paint paint = Paint()
      ..color = Colors.blue.withAlpha(64)
      ..strokeWidth = 4.0
      ..strokeJoin = StrokeJoin.miter
      ..strokeCap = StrokeCap.square;
    canvas.drawRawPoints(mode, list, paint);
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) => true;
}

class PathPainter extends CustomPainter {
  PathPainter(this.list);

  final List<MobilePoint> list;

  @override
  void paint(Canvas canvas, Size size) {
    Paint paint = Paint()
      ..color = Colors.blue.withAlpha(64)
      ..style = PaintingStyle.stroke
      ..strokeWidth = 4.0
      ..strokeJoin = StrokeJoin.miter
      ..strokeCap = StrokeCap.square;
    Path path = Path();
    path.moveTo(list[0].x, list[1].y);
    for (int i = 1; i < list.length; i++) {
      path.lineTo(list[i].x, list[i].y);
    }
    canvas.drawPath(path, paint);
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) => true;
}
```
</details>

### CODE REVIEW

File: impeller/display_list/display_list_dispatcher.cc [hunk 1]
```diff
@@ -762,7 +762,38 @@ void DisplayListDispatcher::drawArc(const SkRect& oval_bounds,
...
+  Paint paint = paint_;
+  paint.style = Paint::Style::kStroke;
+  switch (mode) {
+    case SkCanvas::kPoints_PointMode:
+      if (paint.stroke_cap == SolidStrokeContents::Cap::kButt) {
+        paint.stroke_cap = SolidStrokeContents::Cap::kSquare;
+      }
+      for (uint32_t i = 0; i < count; i++) {
+        SkPoint p0 = points[i];
+        SkPoint p1 = points[i] + SkPoint{0.0001, 0.0};
```
Comment (bdero) [inline impeller/display_list/display_list_dispatcher.cc:775]: Maybe use `kEhCloseEnough` for this hack? 🤣

**Review by bdero — COMMENTED** [review]
Is there supposed to be overdraw happening for a single `drawPoints` call? I suspect the answer is yes but just in case: I used a stencil to [prevent overdraw](https://github.com/flutter/flutter/issues/101330) from happening when drawing strokes internally, so if overdraw isn't supposed to happen, building/drawing a single path with multiple contours can work here (as opposed to building/drawing several single contour paths).

image

**Comment by flar** [comment]
> Is there supposed to be overdraw happening for a single `drawPoints` call? I suspect the answer is yes but just in case: I used a stencil to [prevent overdraw](https://github.com/flutter/flutter/issues/101330) from happening when drawing strokes internally, so if overdraw isn't supposed to happen, building/drawing a single path with multiple contours can work here (as opposed to building/drawing several single contour paths).

I said exactly this in the description. This is what drawPoints does. Creating a single path will produce the wrong result, though there might be more optimal implementations that would put all of the lines into a single op and inform the stroking code NOT to do its over-draw optimizations, but that would involve a lot of digging around in the internals. Meanwhile, this dozen or so lines of code does exactly what the op is supposed to do.

**Review by bdero — APPROVED** [review]

File: impeller/display_list/display_list_dispatcher.cc (hunk 1 above)
Comment (flar) [reply impeller/display_list/display_list_dispatcher.cc]: Done. See https://github.com/flutter/flutter/issues/109077 and the new comment about `kEhCloseEnough`

**Comment by bdero** [comment]
Sounds good and I agree: If it matches Skia's behavior we should land it.

**Review by bdero — APPROVED** [review]

**Pull request merged**